### ISSUE TITLE
Supporting bundled java

### ISSUE THREAD

**Issue by oxodao** [SKCraft/Launcher#521]
I know this would be a lot of work but this could really ease the deployment on end users. This is not a request but a proposal to discuss about this issue and how we could handle it properly if we were to implement it.

The official launcher, being in electron is native and downloads the correct java version by itself. 

The idea would be that we build a replacement for the bootstrap in another language (native, e.g. golang / c++ / rust or one with native bundlers like python. This can probably be done with GraalVM in Java directly but I don't know much about it). This bootstrap's job would be like the current one to download the real launcher, but also a working Java version for it as there is no real JRE for newer java AFAIK (I had to make my friends install JDK...) 

Once the launcher is self-sufficient, we could assign a JRE version in the modpack's manifest with some kind of API (Maybe with [Adoptium's API](https://api.adoptium.net/q/swagger-ui/#/) so that each modpack can specify which version it should use.

The result would be the enduser having nothing to do to be able to play: just download the bootstrap for its OS, run it and everything gets done for him

This relates to this previous issue: https://github.com/SKCraft/Launcher/issues/423

EDIT: Just found out there was something previously started to do this: https://github.com/SKCraft/rusty-bootstrapper
EDIT2: After poking around with the launcher, I've written the steps used by the official launcher on the wiki.vg: https://wiki.vg/Game_files

**Comment by hedgehog1029** [comment]
sorry about the silence on this! I read the issue, mentally went "yeah sounds good but is a lot of work", and then never replied :sweat_smile: 

As you noted the Rust bootstrapper was started to give a jumping off point for something like this. By shifting the bootstrapper to native we can then start to look at having the launcher bundle its own runtime (ideally using jpackage or similar).

I think that's largely a separate issue from auto-managing game runtimes though; that could be implemented even without the bootstrapper changing. I've kind of shied away from it in the past because I know Oracle licensing is an abyss that you don't want to gaze into, but it might be possible to piggyback off the official launcher runtimes without running afoul of the license. It would make the most sense - each game version is specified to run with a specific runtime that the launcher metadata specifies, so by using that same download we would ensure we're not causing issues in-game by using a slightly different Java build.

**Comment by oxodao** [comment]
Ah yes i didn't think about this but using the official launcher installed java could be nice too so that we can save on space (I have a few friends that play on 128gb macbook that would benefit from this). Though that would require the user to install the versions you want to play on the official launcher before using it in a modpack

**Comment by hedgehog1029** [comment]
Yeah, native bootstrap + repackaging will mean I can upgrade from Java 8. Though I only recently bumped it from Java 6, so there's still quite a lot of old code kicking about.

I don't necessarily mean "just use the official launcher's folders" - the current runtime finder already does that - I just meant "download Java from the same source as the official launcher". We're already downloading libraries from libraries.minecraft.net - it isn't too much of a stretch to download Java runtimes too.
